test/Test/Database/LSMTree/Internal/IndexOrdinary.hs

```diff
@@ -0,0 +1,525 @@
+    = errorMsgOrIndex === noErrorMsgButCorrectIndex
+    where
+
+    errorMsgOrIndex :: Either String IndexOrdinary
+    errorMsgOrIndex = snd <$> fromSBS (serialisedIndex entryCount lastKeys)
+
+    noErrorMsgButCorrectIndex :: Either String IndexOrdinary
+    noErrorMsgButCorrectIndex = Right (indexFromLastKeys lastKeys)
+
+prop_tooShortInputMakesDeserialisationFail :: TooShortByteString -> Bool
+prop_tooShortInputMakesDeserialisationFail
+    = isLeft . fromSBS . fromTooShortByteString
+
+prop_typeAndVersionErrorMakesDeserialisationFail :: Word32
+                                                 -> [SerialisedKey]
+                                                 -> NumEntries
+                                                 -> Property
+prop_typeAndVersionErrorMakesDeserialisationFail typeAndVersion
+                                                 lastKeys
+                                                 entryCount
+    = typeAndVersion /= testedTypeAndVersion ==> isLeft errorMsgOrResult
+    where
+
+    errorMsgOrResult :: Either String (NumEntries, IndexOrdinary)
+    errorMsgOrResult
+        = fromSBS $
+          potentialSerialisedIndex (byteVectorFromPrim typeAndVersion)
+                                   (lastKeysBlocks lastKeys)
+                                   (entryCountBlock entryCount)
+
+prop_partialKeySizeBlockMakesDeserialisationFail :: [SerialisedKey]
+                                                 -> Word8
+                                                 -> NumEntries
+                                                 -> Bool
+prop_partialKeySizeBlockMakesDeserialisationFail lastKeys
+                                                 partialKeySizeByte
+                                                 entryCount
+    = isLeft $
+      fromSBS $
+      potentialSerialisedIndex
+          testedTypeAndVersionBlock
+          (lastKeysBlocks lastKeys ++ [Primitive.singleton partialKeySizeByte])
+          (entryCountBlock entryCount)
+
+prop_partialKeyBlockMakesDeserialisationFail :: [SerialisedKey]
+                                             -> Small Word16
+                                             -> Primitive.Vector Word8
+                                             -> NumEntries
+                                             -> Property
+prop_partialKeyBlockMakesDeserialisationFail lastKeys
+                                             (Small statedSize)
+                                             partialKeyBlock
+                                             entryCount
+    = fromIntegral statedSize > Primitive.length partialKeyBlock ==>
+      isLeft (fromSBS input)
+    where
+
+    statedSizeBlock :: Primitive.Vector Word8
+    statedSizeBlock = byteVectorFromPrim statedSize
+
+    input :: ShortByteString
+    input = potentialSerialisedIndex
+                testedTypeAndVersionBlock
+                (lastKeysBlocks lastKeys ++ [statedSizeBlock, partialKeyBlock])
```

Comment:
Yes, nice. Maybe one more test with a completely random key block would make sense, where we don't care about Left/Right, just that the result is `QC.total`. But this here should already cover the case of a key length going out of range. :+1:

Reply:
I think that a key length being too large is really the only error that can occur in a serialized key list beside the error where the serialized key list ends in a partial key size specification. The keys themselves can be arbitrary sequences of bytes; so, as long as the key length specifications are correct, any sequence of bytes is a valid serialized key list.

Just imagine how you can get the keys out of some given sequence of bytes: you take the size information from the beginning, chop off correspondingly many bytes to get the first key, then take the next size information, chop off correspondingly many bytes to get the second key, and so on. If you have an empty sequence of bytes at the end, everything is fine.

So the bottom line is that *any* random sequence of bytes is already covered by the tests, as it falls into one of the following three cases:

* Valid serialized key list
* Key list with an incomplete key size specification at the end
* Key list with a final key size specification that is too large

Reply:
That is true, as long as the generators for each of these cases cover the full space of inputs for that case. Looks good.